### PR TITLE
rewriter: Define call gates for function pointers in source defining the pointee

### LINKED ISSUE #435: Support compiling with `-fvisibility=hidden`
Split into:
* [x] #442
* [ ] #443

Compiling with `-fvisibility=hidden` gives the following error which @kkysen hit with dav1d. This is caused because the undefined references become local symbols in the other .so so we need explicit visibility attributes on those variables to avoid this.
```
[7/7] Linking C executable tests/two_keys_minimal/two_keys_minimal
FAILED: tests/two_keys_minimal/two_keys_minimal
cd /home/ayrton/immunant/IA2-Phase2/build/tests/two_keys_minimal && /usr/bin/objcopy --redefine-syms=/home/ayrton/immunant/IA2-Phase2/build/tests/two_keys_minimal/two_keys_minimal_call_gates_1.objcopy /home/ayrton/immunant/IA2-Phase2/build/tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/./main.c.o && cd /home/ayrton/immunant/IA2-Phase2/build && /usr/bin/cc  -fsanitize=undefined -Wl,--export-dynamic -Wl,@/home/ayrton/immunant/IA2-Phase2/build/tests/two_keys_minimal/two_keys_minimal_call_gates_1.ld -Wl,--print-map-locals -pthread -Wl,--wrap=pthread_create -Wl,--wrap=main -Wl,-z,now -Wl,-z,relro -Wl,-T/home/ayrton/immunant/IA2-Phase2/build/runtime/libia2/padding.ld -Wl,--dynamic-list=/home/ayrton/immunant/IA2-Phase2/build/runtime/libia2/dynsym.syms -Wl,-wrap,calloc -Wl,-wrap,free -Wl,-wrap,malloc -Wl,-wrap,memalign -Wl,-wrap,posix_memalign -Wl,-wrap,pvalloc -Wl,-wrap,realloc -Wl,-wrap,valloc -Wl,-wrap,malloc_usable_size -Wl,-wrap,realpath -Wl,-wrap,strdup -Wl,-wrap,strndup -Wl,-wrap,getcwd -Wl,-wrap,asprintf -Wl,-wrap,vasprintf tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o -o tests/two_keys_minimal/two_keys_minimal  -Wl,-rpath,/home/ayrton/immunant/IA2-Phase2/build/runtime/partition-alloc:/home/ayrton/immunant/IA2-Phase2/build/tests/two_keys_minimal:/home/ayrton/immunant/IA2-Phase2/build/runtime/libia2  -ldl  runtime/libia2/liblibia2.a  runtime/partition-alloc/libpartition-alloc.so  -lcriterion  tests/two_keys_minimal/libtwo_keys_minimal_lib.so  tests/two_keys_minimal/libtwo_keys_minimal_call_gates.so  runtime/libia2/liblibia2.a  -ldl  runtime/libia2/libc.so.6 && :
/usr/bin/ld: tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o: in function `ia2_stackptr_for_pkru':
main.c:(.text+0x24e): undefined reference to `ia2_stackptr_2'
/usr/bin/ld: tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o: in function `init_stacks_and_setup_tls':
main.c:(.text+0x2bf): undefined reference to `init_tls_2'
/usr/bin/ld: main.c:(.text+0x2d0): undefined reference to `ia2_stackptr_2'
/usr/bin/ld: tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o: in function `ia2_init':
main.c:(.text+0x395): undefined reference to `ia2_setup_destructors_2'
/usr/bin/ld: tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o: in function `print_message':
main.c:(.text+0x17fd): undefined reference to `plugin_secret'
/usr/bin/ld: tests/two_keys_minimal/two_keys_minimal: hidden symbol `two_keys_plugin_impl' in tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o is referenced by DSO
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

### PR DESCRIPTION
**Pull request by ayrtonm**
We previously defined call gates for pointers to static functions by appending a macro to the rewritten source files. In that case we had to do this because the static function may not be referenced from outside the translation unit. For pointers to globally-visible functions we defined the call gates in the call gate .so for simplicity. This commit changes the latter call gates so they're defined like the former. This change simplifies control-flow since the call gate no longer requires cross-DSO calls and makes it possible to support programs that are intended to be compiled with -fvisibility=hidden since functions that may have hidden visibility are only referenced by call gates within the DSO that defines them. Partially addresses #435.

### CODE REVIEW

**Comment by ayrtonm** [comment]
Turns out we still need to distinguish between static and non-static address-taken functions even if they're both defined in a compartment .c because in the non-static case the call gate's target function may also be called directly from another compartment. When that happens the `-Wl,--wrap=some_function` we pass for the direct call changes the function that the indirect call gate references. To fix this I need to explicitly use `__real_some_function` in the indirect call gates.

**Comment by ayrtonm** [comment]
I also needed to distinguish between call gates for non-static address-taken functions in the same DSO and those for functions in another DSO. PR needs cleanup and updating comments but it should be working now.

**Comment by ayrtonm** [comment]
TODO: check `should_not_modify_file` before adding the used attribute and create a map between input and output files in `copy_files`.

File: tools/rewriter/SourceRewriter.cpp [hunk 1]
```diff
@@ -1435,10 +1450,18 @@ int main(int argc, const char **argv) {
...
+                           true /* as_macro */);
+      macros_defining_wrappers += "#define IA2_DEFINE_WRAPPER_"s + fn_name + " \\\n";
+      macros_defining_wrappers += "asm(\\\n";
+      macros_defining_wrappers += asm_wrapper;
+      macros_defining_wrappers += ");\n";
+
+      /* Invoke the macro we just defined in the source file defining the target function */
+      auto filename = fn_decl_pass.fn_definitions[fn_name];
+      std::ofstream source_file(filename, std::ios::app);
+      source_file << "IA2_DEFINE_WRAPPER(" << fn_name << ")\n";
```
Comment (kkysen) [inline tools/rewriter/SourceRewriter.cpp:1463]: Could we document how `IA2_DEFINE_WRAPPER()` and `IA2_DEFINE_WRAPPER_*` work somewhere?

**Review by kkysen — COMMENTED** [review]
When I try this on `dav1d`, I'm able to successfully compile with `-fvisibility=hidden` again.  The `DEFINE_WRAPPER(...)`s appear to be working.  However, at runtime, I'm getting symbol lookup errors for these symbols:
* `dav1d_num_logical_processors`
* `dav1d_init_cpu`
* `dav1d_get_cpu_flags_x86`

These symbols are all marked `COLD` (`__attribute__((cold))`), but when I tried removing this, the error persisted.  I tried `__attribute__((used))`, too, which didn't work either.  Adding `DAV1D_API` (`__attribute__((visibility ("default")))`) did work, but I'm not really sure why this is needed.

File: tools/rewriter/SourceRewriter.cpp [hunk 2]
```diff
@@ -920,6 +922,7 @@ class FnDecl : public RefactoringCallback {
   std::set<Function> declared_fns[MAX_PKEYS];
   std::map<Function, AbiSignature> abi_signatures;
   std::map<Function, Pkey> fn_pkeys;
+  std::map<Function, Filename> fn_definitions;
```
[the remark this answers is not in the document]
Comment (ayrtonm) [reply tools/rewriter/SourceRewriter.cpp:925]: it shouldn't need to be

File: tools/rewriter/SourceRewriter.cpp [hunk 3]
```diff
@@ -1418,8 +1421,20 @@ int main(int argc, const char **argv) {
   }
 
   std::cout << "Generating function pointer wrappers\n";
+  std::string macros_defining_wrappers;
+  /*
+   * This loops over all non-static address-taken functions but we only want to
+   * define one call gate for each so we need to track them in case a function
+   * has its address taken in multiple places
+   */
+  std::set<Function> generated_wrappers = {};
```
[the remark this answers is not in the document]
Comment (ayrtonm) [reply tools/rewriter/SourceRewriter.cpp:1430]: it shouldn't need to be

File: tools/rewriter/SourceRewriter.cpp (hunk 1 above)
Comment (ayrtonm) [reply tools/rewriter/SourceRewriter.cpp:1467]: The comment was kind of misleading so I clarified it. This is one of those places where we're still trying to figure out what's the best way to interpose call gates as we get feedback from applying it to codebases so I don't think writing excessive documentation for things that may soon be out of date is a good use of our time.

**Pull request merged**